### PR TITLE
A fix & some refactoring changes.

### PR DESCRIPTION
**Pull request by Martinski4GitHub**
- Fixed the case where the Main Menu was showing the "Force Update Script Now" option but was not actually updating when selected.
- Some refactoring changes & fine-tuning code.

### CODE REVIEW

**Comment by Martinski4GitHub** [comment]
@ExtremeFiretop,

As always, please review & test the PR whenever you have the time.
I don't think this new feature is critical functionality so we can take our time to get it out the door.

**Comment by ExtremeFiretop** [comment]
I'm currently watching some Plex with the girlfriend, but I'll review this first thing! Happy July 4th of I don't talk to you until tomorrow! (I know I'm probably about an hour and a bit early)

**Comment by Martinski4GitHub** [comment]
> I'm currently watching some Plex with the girlfriend, but I'll review this first thing! Happy July 4th if I don't talk to you until tomorrow! (I know I'm probably about an hour and a bit early)
>
>
LOL!!  Here, we're about to continue binge-watching the Amazon series we started last night. LOL!!

So no worries. We can take our time on this one. Nothing critical.

Enjoy your movie or whatever you're watching & I'll do the same. We deserve some time off!!

**Comment by ExtremeFiretop** [comment]
> > I'm currently watching some Plex with the girlfriend, but I'll review this first thing! Happy July 4th if I don't talk to you until tomorrow! (I know I'm probably about an hour and a bit early)
> 
> LOL!! Here, we're about to continue binge-watching the Amazon series we started last night. LOL!!
> 
> So no worries. We can take our time on this one. Nothing critical.
> 
> Enjoy your movie or whatever you're watching & I'll do the same. We deserve some time off!!

Sounds good buddy! We are watching some TV, right now it's some intervention LOL!

I'll be up tomorrow and merge this in in the morning after testing. 😉

File: MerlinAU.sh [hunk 1]
```diff
@@ -683,55 +690,50 @@ _SCRIPTUPDATE_()
...
-		SCRIPT_URL_BASE="https://raw.githubusercontent.com/ExtremeFiretop/MerlinAutoUpdate-Router/$SCRIPT_BRANCH"
-		_SCRIPTUPDATE_ force
-		_DoExit_ 0
+##----------------------------------------##
+## Modified by Martinski W. [2024-Jul-03] ##
+##----------------------------------------##
+_ChangeToDev_()
+{
+    SCRIPT_BRANCH="dev"
+    SCRIPT_URL_REPO="${SCRIPT_URL_BASE}/$SCRIPT_BRANCH"
```
Comment (ExtremeFiretop) [inline MerlinAU.sh:723]: Approved.

File: MerlinAU.sh [hunk 2]
```diff
@@ -11,12 +11,15 @@ set -u
...
+##----------------------------------------##
+## Modified by Martinski W. [2024-Jul-03] ##
+##----------------------------------------##
 # Script URL Info #
-SCRIPT_BRANCH="master"
-SCRIPT_URL_BASE="https://raw.githubusercontent.com/ExtremeFiretop/MerlinAutoUpdate-Router/$SCRIPT_BRANCH"
+
+## Set to "master" for Production Releases ##
+SCRIPT_BRANCH="dev"
+readonly SCRIPT_URL_BASE="https://raw.githubusercontent.com/ExtremeFiretop/MerlinAutoUpdate-Router"
```
Comment (ExtremeFiretop) [inline MerlinAU.sh:21]: Approved.

File: MerlinAU.sh [hunk 3]
```diff
@@ -250,9 +253,10 @@ logo() {
 ##---------------------------------------##
 ## Added by ExtremeFiretop [2024-Jul-03] ##
 ##---------------------------------------##
-_ShowAbout_(){
+_ShowAbout_()
+{
     logo
-	cat <<EOF
+    cat <<EOF
```
Comment (ExtremeFiretop) [inline MerlinAU.sh:259]: Approved

File: MerlinAU.sh [hunk 4]
```diff
@@ -267,16 +271,17 @@ Help & Support
...
 
 ##---------------------------------------##
 ## Added by ExtremeFiretop [2024-Jul-03] ##
 ##---------------------------------------##
-_ShowHelp_(){
+_ShowHelp_()
+{
     logo
-	cat <<EOF
+    cat <<EOF
```
Comment (ExtremeFiretop) [inline MerlinAU.sh:284]: Approved

File: MerlinAU.sh [hunk 5]
```diff
@@ -579,7 +584,7 @@ _CheckForNewScriptUpdates_()
    rm -f "$SCRIPTVERPATH"
 
    # Download the latest version file from the source repository
-   curl -LSs --retry 4 --retry-delay 5 "${SCRIPT_URL_BASE}/version.txt" -o "$SCRIPTVERPATH"
+   curl -LSs --retry 4 --retry-delay 5 "${SCRIPT_URL_REPO}/version.txt" -o "$SCRIPTVERPATH"
```
Comment (ExtremeFiretop) [inline MerlinAU.sh:587]: Approved

File: MerlinAU.sh [hunk 6]
```diff
@@ -683,55 +690,50 @@ _SCRIPTUPDATE_()
...
-		SCRIPT_URL_BASE="https://raw.githubusercontent.com/ExtremeFiretop/MerlinAutoUpdate-Router/$SCRIPT_BRANCH"
-		_SCRIPTUPDATE_ force
-		_DoExit_ 0
+##----------------------------------------##
+## Modified by Martinski W. [2024-Jul-03] ##
+##----------------------------------------##
+_ChangeToStable_()
+{
+    SCRIPT_BRANCH="master"
+    SCRIPT_URL_REPO="${SCRIPT_URL_BASE}/$SCRIPT_BRANCH"
```
Comment (ExtremeFiretop) [inline MerlinAU.sh:734]: Approved

File: MerlinAU.sh [hunk 7]
```diff
@@ -5870,11 +5872,13 @@ _ShowMainMenu_()
...
    if [ "$scriptUpdateNotify" != "0" ]
    then
+      forceScriptUpdateOption=""
       printf "\n ${GRNct}up${NOct}.  Update $SCRIPT_NAME Script Now"
-      printf "\n${padStr}[Version: ${GRNct}${DLRepoVersion}${NOct} Available for Download]\n"
+      printf "\n${padStr}[Version ${GRNct}${DLRepoVersion}${NOct} Available for Download]\n"
    else
-      printf "\n ${GRNct}up${NOct}.  Force Update $SCRIPT_NAME Now"
-      printf "\n${padStr}[No Updates Available.]\n"
+      forceScriptUpdateOption="force"
```
Comment (ExtremeFiretop) [inline MerlinAU.sh:5879]: Approved

File: MerlinAU.sh [hunk 8]
```diff
@@ -683,55 +690,50 @@ _SCRIPTUPDATE_()
...
       if _WaitForYESorNO_
       then
-          echo ; echo
+          printf "\n\n"
           echo -e "${CYANct}Downloading $SCRIPT_NAME ${CYANct}v${DLRepoVersion}${NOct}"
-          curl -LSs --retry 4 --retry-delay 5 "${SCRIPT_URL_BASE}/version.txt" -o "$SCRIPTVERPATH"
-          curl -LSs --retry 4 --retry-delay 5 "${SCRIPT_URL_BASE}/${SCRIPT_NAME}.sh" -o "$ScriptFileDL"
 
-          if [ $? -eq 0 ] && [ -s "$ScriptFileDL" ]
+          if _DownloadScriptFiles_
```
Comment (ExtremeFiretop) [inline MerlinAU.sh:696]: Approved

File: MerlinAU.sh [hunk 9]
```diff
@@ -683,55 +690,50 @@ _SCRIPTUPDATE_()
...
-              rm -f "$ScriptFileDL"
-              echo
-              echo -e "${REDct}Download failed.${NOct}"
               _WaitForEnterKey_
               return
           fi
       else
-          echo ; echo
-          echo -e "${GRNct}Exiting Update Utility...${NOct}"
+          printf "\n\n"
```
Comment (ExtremeFiretop) [inline MerlinAU.sh:709]: Approved

File: MerlinAU.sh [hunk 10]
```diff
@@ -606,75 +611,77 @@ ${REDct}v${SCRIPT_VERSION}${NOct} --> ${GRNct}v${DLRepoVersion}${NOct}"
...
       if _WaitForYESorNO_
       then
-          echo ; echo
+          printf "\n\n"
           echo -e "${CYANct}Downloading $SCRIPT_NAME ${CYANct}v${DLRepoVersion}${NOct}"
-          curl -LSs --retry 4 --retry-delay 5 "${SCRIPT_URL_BASE}/version.txt" -o "$SCRIPTVERPATH"
-          curl -LSs --retry 4 --retry-delay 5 "${SCRIPT_URL_BASE}/${SCRIPT_NAME}.sh" -o "$ScriptFileDL"
 
-          if [ $? -eq 0 ] && [ -s "$ScriptFileDL" ]
+          if _DownloadScriptFiles_
```
Comment (ExtremeFiretop) [inline MerlinAU.sh:673]: Approved

File: MerlinAU.sh [hunk 11]
```diff
@@ -606,75 +611,77 @@ ${REDct}v${SCRIPT_VERSION}${NOct} --> ${GRNct}v${DLRepoVersion}${NOct}"
...
-            _ReleaseLock_
-            exec "$ScriptFilePath"
-        else
-            rm -f "$ScriptFileDL"
-            echo -e "${REDct}Download failed.${NOct}"
-        fi
-         return
-    fi
+
+   _DownloadScriptFiles_()
```
Comment (ExtremeFiretop) [inline MerlinAU.sh:623]: Approved

File: MerlinAU.sh [hunk 12]
```diff
@@ -606,75 +611,77 @@ ${REDct}v${SCRIPT_VERSION}${NOct} --> ${GRNct}v${DLRepoVersion}${NOct}"
...
    clear
    logo
-   echo
-   echo -e "${YLWct}Update Utility${NOct}"
-   echo
-   echo -e "${CYANct}Current Version: ${YLWct}${SCRIPT_VERSION}${NOct}"
-   echo -e "${CYANct}Updated Version: ${YLWct}${DLRepoVersion}${NOct}"
-   echo
+   printf "\n${YLWct}Script Update Utility${NOct}\n\n"
+   printf "${CYANct}Version Currently Installed:  ${YLWct}${SCRIPT_VERSION}${NOct}\n"
```
Comment (ExtremeFiretop) [inline MerlinAU.sh:661]: Approved

File: MerlinAU.sh [hunk 13]
```diff
@@ -6204,7 +6210,7 @@ do
           else _Approve_FW_Update_
           fi
           ;;
-      up) _SCRIPTUPDATE_
+      up) _SCRIPTUPDATE_ "$forceScriptUpdateOption"
```
Comment (ExtremeFiretop) [inline MerlinAU.sh:6213]: Ah, I see the bug. 
I forgot to include the optional "force" for this.

**Review by ExtremeFiretop — APPROVED** [review]

File: MerlinAU.sh (hunk 13 above)
Comment (ExtremeFiretop) [reply MerlinAU.sh:6213]: Actually, this is not the exact design I intended.
I'll submit another PR with the suggested changes of the original design choices I intended to do.

Let me know what you prefer.

Comment (ExtremeFiretop) [reply MerlinAU.sh:6213]: So my intention was not to use the force here actually.
It was to use the regular update function. 

The bug, was actually in my _SCRIPTUPDATE_ function when setting: `if [ "$1" = "force" ]`

File: MerlinAU.sh [hunk 14]
```diff
@@ -606,75 +611,77 @@ ${REDct}v${SCRIPT_VERSION}${NOct} --> ${GRNct}v${DLRepoVersion}${NOct}"
...
+          retCode=0
+      else
+          rm -f "$ScriptFileDL"
+          printf "\n${REDct}Download failed.${NOct}\n"
+          retCode=1
+      fi
+      return "$retCode"
+   }
+
+   if [ $# -gt 0 ] && [ "$1" = "force" ]
```
Comment (ExtremeFiretop) [inline MerlinAU.sh:643]: Adjusting this `if [ "$1" = "force" ]` to this: `if [ $# -gt 0 ] && [ "$1" = "force" ]`
Is what ultimately resolved the unset $1 bug, but I didn't intend on using the force in the menu ever, just the regular update function.

**Comment by ExtremeFiretop** [comment]
Synced with Gnuton here: https://github.com/ExtremeFiretop/MerlinAutoUpdate-Router/pull/186/commits/a3dd4551e6c9f3f9731d91396b0000f6e539bfc7

**Pull request merged**